extension/sigv4authextension/extension.go

```diff
@@ -0,0 +1,181 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sigv4authextension // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/sigv4authextension"
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	v4 "github.com/aws/aws-sdk-go/aws/signer/v4"
+	"go.opentelemetry.io/collector/component/componenthelper"
+	"go.opentelemetry.io/collector/config/configauth"
+	"go.uber.org/zap"
+	grpcCredentials "google.golang.org/grpc/credentials"
+)
+
+// Sigv4Auth is a struct that implements the configauth.ClientAuthenticator interface. 
+// It provides the implementation for providing Sigv4 authentication for HTTP requests only.
+type Sigv4Auth struct {
+	cfg 				*Config
+	logger 				*zap.Logger
+	runtimeInfo 			string
+	componenthelper.StartFunc	// embedded default behavior to do nothing with Start()
+	componenthelper.ShutdownFunc	// embedded default behavior to do nothing with Shutdown()
+}
+
+// compile time check that the Sigv4Auth struct satisfies the configauth.ClientAuthenticator interface
+var _ configauth.ClientAuthenticator = (*Sigv4Auth)(nil)
+
+// SigningRoundTripper is a custom RoundTripper that performs AWS Sigv4.
+type SigningRoundTripper struct {
+	transport 	http.RoundTripper
+	signer 		*v4.Signer
+	region 		string
+	service 	string
+	runtimeInfo 	string
+}
+
+// RoundTripper() returns a custom SigningRoundTripper.
+func (sa *Sigv4Auth) RoundTripper(base http.RoundTripper) (http.RoundTripper, error) {
+	cfg := sa.cfg
+
+	creds, err := getCredsFromConfig(cfg)
```

Comment:
I think this should be called as part of `Config.Validate()` to give quick feedback on bad credentials (you can store the creds into a private field on the Config). Or otherwise probably still should be in the extension factory codepath and not here, which will mysteriously fail when initializing a different component (e.g. prometheus exporter), not the extension, I think.

Reply:
Fixed the duplicate call here and removed the `RoundTripperWithCredentials()` method.